dax/internal/client/cluster.go

```diff
@@ -80,344 +85,289 @@ type connConfig struct {
-		req.Error = err
-		return
-	}
-	action := func(client DaxAPI, o RequestOptions) error {
-		o.applyTo(req)
-		client.send(req)
-		return req.Error
-	}
-	if err := cc.retry(req.Operation.Name, action, opt); err != nil {
-		req.Error = err
-	}
-}
-
-func (cc *ClusterDaxClient) retry(op string, action func(client DaxAPI, o RequestOptions) error, opt RequestOptions) (err error) {
+func (cc *ClusterDaxClient) retry(ctx context.Context, op string, action func(client DaxAPI, o RequestOptions) error, opt RequestOptions) (err error) {
 	defer func() {
 		if daxErr, ok := err.(daxError); ok {
 			err = convertDaxError(daxErr)
 		}
 	}()
 
-	ctx := cc.newContext(opt)
-
-	var sleepFun func() error
-	if opt.RetryDelay > 0 {
-		retryDelay := opt.RetryDelay
-		if opt.SleepDelayFn == nil {
-			sleepFun = func() error {
-				return aws.SleepWithContext(ctx, retryDelay)
-			}
-		} else {
-			sleepFun = func() error {
-				opt.SleepDelayFn(retryDelay)
-				return nil
-			}
-		}
-	}
-
-	attempts := opt.MaxRetries
-	opt.MaxRetries = 0 // disable retries on single node client
+	attempts := opt.RetryMaxAttempts
+	opt.RetryMaxAttempts = 0 // disable retries on single node client
 
-	var req request.Request
-	var ok bool
 	var client DaxAPI
 	// Start from 0 to accomodate for the initial request
 	for i := 0; i <= attempts; i++ {
-		if i > 0 && opt.Logger != nil && opt.LogLevel.Matches(aws.LogDebugWithRequestRetries) {
-			opt.Logger.Log(fmt.Sprintf("DEBUG: Retrying Request %s/%s, attempt %d", service, op, i))
+		if i > 0 && opt.Logger != nil {
+			opt.Logger.Logf(ClassificationDebug, "Retrying Request %s/%s, attempt %d", service, op, i)
 		}
-		client, err = cc.cluster.client(client)
+		client, err = cc.cluster.client(client, op)
 		if err != nil {
-			if req, ok = cc.shouldRetry(opt, err); !ok {
+			if !isRetryable(opt, i+1, err) {
 				return err
 			}
+			continue
 		}
 
+		err = action(client, opt)
 		if err == nil {
-			if err = action(client, opt); err == nil {
-				return nil
-			} else if req, ok = cc.shouldRetry(opt, err); !ok {
-				return err
-			}
+			// success
+			return nil
+		}
+		if !isRetryable(opt, i+1, err) {
+			return err
 		}
 
-		if i != attempts {
-			req.RetryCount = i + 1
-			delay := opt.Retryer.RetryRules(&req)
-			if delay != 0 {
-				if opt.SleepDelayFn == nil {
-					aws.SleepWithContext(ctx, delay)
-				} else {
-					opt.SleepDelayFn(delay)
-				}
-			} else if sleepFun != nil {
-				if err := sleepFun(); err != nil {
-					return awserr.New(request.CanceledErrorCode, "request context canceled", err)
-				}
-			}
+		d, err := opt.Retryer.RetryDelay(i+1, err)
+		if err != nil {
+			return &smithy.OperationError{Err: err, OperationName: op}
+		}
+		if err = Sleep(ctx, op, d); err != nil {
+			return err
+		}
 
-			if err != nil && opt.Logger != nil && opt.LogLevel.Matches(aws.LogDebugWithRequestRetries) {
-				opt.Logger.Log(fmt.Sprintf("DEBUG: Error in executing request %s/%s. : %s", service, op, err))
-			}
+		if opt.Logger != nil {
```

Comment:
here and in other places, why do you need this check? set a default `NoOp` logger :)